### PR TITLE
Unpin `pip` version and bump actions versions

### PR DESCRIPTION
**Pull request by ankatiyar**
## Description

Partially fix https://github.com/kedro-org/kedro/issues/3184
(Also needs to be updated on `kedro-plugins` in a separate PR)

`pip` released a new version -> https://pypi.org/project/pip/24.0/

## Development notes

* Update versions of checkout, setup-python, and setup-msbuild actions
Screenshot 2024-02-05 at 17 23 53

* unpin `pip` version

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Comment by astrojuanlu** [comment]
🤞🏽

File: .github/workflows/auto-merge-prs.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
```
Comment (noklam) [inline .github/workflows/auto-merge-prs.yml:15]: What does this change?

**Review by noklam — APPROVED** [review]
I've gone through the [Release note of pip ](https://pip.pypa.io/en/stable/news/)

> Normalize extras according to [PEP 685](https://peps.python.org/pep-0685/) from package metadata in the resolver for comparison. This ensures extras are correctly compared and merged as long as the package providing the extra(s) is built with values normalized according to the standard. Note, however, that this does not solve cases where the package itself contains unnormalized extra values in the metadata. ([#11649](https://github.com/pypa/pip/issues/11649))

Noted the original problem was related to the unnormalised extra names we have in `kedro-datasets`. I suspect the tests passed here because we don't have dataset extras in `kedro` anymore. So I suggest when you try to make the same changes in `kedro-datasets`, make sure things like `kedro-datasets[all]` will work as expected, right now I think it should fail.

https://peps.python.org/pep-0685/

File: .github/workflows/auto-merge-prs.yml (hunk 1 above)
Comment (ankatiyar) [reply .github/workflows/auto-merge-prs.yml:15]: So the node 16 actions have been deprecated by Github Actions and the new versions of these actions have transitioned to node 20 (https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/)

**Comment by ankatiyar** [comment]
> I've gone through the [Release note of pip ](https://pip.pypa.io/en/stable/news/)
> 
> > Normalize extras according to [PEP 685](https://peps.python.org/pep-0685/) from package metadata in the resolver for comparison. This ensures extras are correctly compared and merged as long as the package providing the extra(s) is built with values normalized according to the standard. Note, however, that this does not solve cases where the package itself contains unnormalized extra values in the metadata. ([#11649](https://github.com/pypa/pip/issues/11649))
> 
> Noted the original problem was related to the unnormalised extra names we have in `kedro-datasets`. I suspect the tests passed here because we don't have dataset extras in `kedro` anymore. So I suggest when you try to make the same changes in `kedro-datasets`, make sure things like `kedro-datasets[all]` will work as expected, right now I think it will most likely fail.
> 
> https://peps.python.org/pep-0685/

I'm trying it out on `kedro-datasets` but i'll leave the ticket open if it doesn't work out.

**Pull request merged**